Feature/meshrendering2

AnkeAnke opened the pull request:
**Rasterizers and Renderers**
We now split the fragment list rendering from a monolithic fancy renderer into two stages:
- Rasterizer processors for different kinds of primitives emit rasterizations, functors that generate fragments
- A renderer sets up the fragment lists, calls the rasterizations, and then evaluates the fragments in each pixel

This allows us to mix settings and primitives by using several rasterizers.

**To be done**
- More primitive renderers (lines, points, spheres, raycasts)
  - Example workspaces
  - Testing
- Mixed opacity settings in rasterizers
  - Background image depth is taken into account now
  - Opaque rasterizers are be ignored for fragment list evaluation, need another in-between render target
- Rasterizations should not hold properties, but only the settings of the generating processor

**What environment has the code been compiled and tested on?**
- Pop! OS (Ubuntu) 19.10
- GCC
- Qt 5.14.1 / CMake 3.13.4 / Python 3.7.5
![CowsPlosion](https://user-images.githubusercontent.com/2829051/78561923-cc476280-7818-11ea-84c0-2c2ffa53b6d6.png)

AnkeAnke committed:
Split fancy rendering via rasterization functors

AnkeAnke committed:
Removing dependent glsl extension

AnkeAnke committed:
Fixes

AnkeAnke committed:
Removing fancy mesh renderer

AnkeAnke committed:
Adding depth tests

AnkeAnke committed:
Better sorting, illumination reads depth

AnkeAnke committed:
Removing workspaces with fancymeshrenderer

AnkeAnke committed:
Tiny fix

AnkeAnke committed:
Fixing non-virtual destruction

petersteneteg committed:
Jenkins: Warning fix

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/inviwo/inviwo?pullRequest=915) <br/>All committers have signed the CLA.

AnkeAnke committed:
Switched optional port to right

AnkeAnke committed:
Fixing port copyright

petersteneteg commented on modules/meshrenderinggl/CMakeLists.txt:
```diff
@@ -1,67 +1,71 @@
+#--------------------------------------------------------------------
+# Add source files
+set(SOURCE_FILES
+    src/datastructures/halfedges.cpp
```
we want to have 1 .cpp for every .h even of all it does is include it's cpp. Main reason is to make sure each .h can compile in isolation.

petersteneteg commented on modules/meshrenderinggl/glsl/oit/display.frag:
```diff
@@ -80,24 +87,29 @@ void main() {
+        abufferPixel unpackedFragment = uncompressPixelData(nextFragment);
+        gl_FragDepth = min(backgroundDepth, unpackedFragment.depth);
+
+        while (unpackedFragment.depth >= 0 && unpackedFragment.depth <= backgroundDepth) {
```
do while?

AnkeAnke replied:
We need to fetch one more fragment than we add in to color. So either we do one extra fetch before the loop as now or check the condition inside the loop again. I think this solution is the cleaner one. Also, the first check is with last depth = 0 here explicitly, that would become quite obscure otherways.

petersteneteg commented on modules/meshrenderinggl/glsl/oit/display.frag:
```diff
@@ -80,24 +87,29 @@ void main() {
-        }
-        gl_FragDepth = depth;
+
         FragData0 = color;
```
where is bgColor added?

AnkeAnke replied:
Blending is set to alpha, so the fragment ops take care of that for us. Since we got the color texture bound as a side effect of binding depth, we could of course blend here ourselves, but I don't see an advantage of that, it's just more code.

petersteneteg commented on modules/meshrenderinggl/include/modules/meshrenderinggl/datastructures/rasterization.h:
```diff
@@ -0,0 +1,67 @@
+
+class IVW_MODULE_MESHRENDERINGGL_API Rasterization {
+public:
+    virtual ~Rasterization() {}
```
= default

petersteneteg commented on modules/meshrenderinggl/include/modules/meshrenderinggl/datastructures/rasterization.h:
```diff
@@ -0,0 +1,67 @@
+    virtual ~Rasterization() {}
+    virtual void rasterize(std::function<void(Shader&)> setUniforms) const = 0;
+    virtual void setImageSize(const ivec2& size) const = 0;
+    virtual bool usesFragmentLists() const = 0;
```
Should add some comment for these functions

petersteneteg replied:
could have a virtual function for the docs to...

petersteneteg commented on modules/meshrenderinggl/include/modules/meshrenderinggl/ports/rasterizationport.h:
```diff
@@ -0,0 +1,44 @@
+
+namespace inviwo {
+
+// using RasterizationInport = DataInport<Rasterization>;
```
remove

petersteneteg commented on modules/meshrenderinggl/include/modules/meshrenderinggl/processors/meshrasterizer.h:
```diff
@@ -1,306 +1,336 @@
+    };
+
+    std::array<FaceSettings, 2> faceSettings_;
+
```
remove empty lines

petersteneteg commented on modules/meshrenderinggl/include/modules/meshrenderinggl/processors/meshrasterizer.h:
```diff
@@ -1,306 +1,336 @@
+        const std::vector<std::shared_ptr<const Mesh>> meshes_;
+        std::vector<std::shared_ptr<const Mesh>> enhancedMeshes_;
+
+        const Camera* camera_;
```
unique ptr

petersteneteg commented on modules/meshrenderinggl/include/modules/meshrenderinggl/processors/meshrasterizer.h:
```diff
@@ -1,306 +1,336 @@
+        virtual void setImageSize(const ivec2& size) const override { imageSize_ = size; }
+        virtual bool usesFragmentLists() const override {
+            return !forceOpaque_ && FragmentListRenderer::supportsFragmentLists();
+        }
```
we generally put this at the top

petersteneteg commented on modules/meshrenderinggl/include/modules/meshrenderinggl/processors/rasterizationrenderer.h:
```diff
@@ -0,0 +1,171 @@
+    IllustrationSettings illustrationSettings_;
+
+    FragmentListRenderer flr_;
+    typename Dispatcher<void()>::Handle flrReload_;
```
use using

AnkeAnke replied:
I fell for that btw: It's typename, not typedef...

petersteneteg replied:
ahh... can't read...

petersteneteg commented on modules/meshrenderinggl/include/modules/meshrenderinggl/processors/meshrasterizer.h:
```diff
@@ -1,306 +1,336 @@
+    /**
+     * \brief Functor object that will render into a fragment list.
+     */
+    class MeshRasterization : public Rasterization {
```
I would make this freestanding instead of being a inside meshrasterizer

petersteneteg commented on modules/meshrenderinggl/src/processors/meshrasterizer.cpp:
```diff
@@ -1,549 +1,523 @@
+    shader_->build();
+}
+
+void MeshRasterizer::MeshRasterization::rasterize(
```
why not pass in the image size here?

petersteneteg commented on modules/meshrenderinggl/src/processors/meshrasterizer.cpp:
```diff
@@ -1,549 +1,523 @@
+        default:
+            lightPos = lightPosition_;
+    }
+    shader_->setUniform("lighting.position", lightPos);
```
could we not just have the MeshRastersize set all these uniform directly on the shader? why pass them along and then set them, when we could do it directly? we only need to set them when they change or we recompile right?

AnkeAnke replied:
That's a very good point and got rid of a lot of code. Thanks!

petersteneteg reviewed (changes requested):
Looks good I think, some minor comment.

What did you do to make all the diffs so huge, github seems to think you changes all lines more or less...

AnkeAnke committed:
Adding sources for each header

AnkeAnke committed:
Small fixes

AnkeAnke committed:
Virtual function for rasterization documentation

petersteneteg commented on modules/meshrenderinggl/include/modules/meshrenderinggl/datastructures/rasterization.h:
```diff
@@ -45,10 +45,32 @@ namespace inviwo {
+     * \brief Get data description for the network interface.
+     * @return Specific information about this rasterization type/instance.
+     */
+    virtual std::string getDescription() const { return "Rasteriation functor."; }
```
I would have this return the Document

AnkeAnke committed:
Moving code

petersteneteg commented on modules/meshrenderinggl/src/processors/meshrasterizer.cpp:
```diff
@@ -523,7 +523,7 @@ void MeshRasterizer::MeshRasterization::update(const MeshRasterizer& rasterizerP
     std::stringstream ss;
     ss << "Mesh rasterization functor with " << enhancedMeshes_.size();
-    ss << (enhancedMeshes_.size() == 1) ? "mesh." : " meshes.";
+    ss << ((enhancedMeshes_.size() == 1) ? " mesh." : " meshes.");
```
fmt::format is quite nice here...

AnkeAnke committed:
Fixing bad fix by actually reading

AnkeAnke committed:
Cleaning up MeshRasterization
Moving outside processor.
Removing uniforms, keeping cull settings and tranfer function texture.

AnkeAnke committed:
Rasterizations returning info document

ResearchDaniel commented on modules/meshrenderinggl/glsl/oit/display.frag:
```diff
@@ -107,7 +119,7 @@ void main() {
         discard;
 #else
-        FragData0 = vec4(0.0f);
+        FragData0 = vec4(1.0f);
```
Why white color? Has this been for debugging? Otherwise I think it should be stated why

ResearchDaniel commented on modules/meshrenderinggl/include/modules/meshrenderinggl/processors/meshrasterizer.h:
```diff
@@ -1,306 +1,315 @@
+        FloatProperty normalVariationExponent_;
+        // Smoke surfaces
+        BoolProperty enableDensity_;
+        FloatProperty baseDensity_;  // k in the paper
```
Which paper? 
Can state that in the AlphaSettings doc?

AnkeAnke replied:
Now citing both papers in the top, since they are mentioned repeatedly

ResearchDaniel commented on modules/meshrenderinggl/include/modules/meshrenderinggl/processors/meshrasterizer.h:
```diff
@@ -1,306 +1,315 @@
+    enum class HatchingMode : char { U, V, UV };
+    enum class HatchingBlendingMode : char { Multiplicative, Additive };
+    /**
+     * \brief Hatching settings. These are exactly the parameters from the IRIS-paper
```
IRIS-paper? Quick google IRIS + hatching did not reveal it...

ResearchDaniel commented on modules/meshrenderinggl/src/processors/rasterizationrenderer.cpp:
```diff
@@ -0,0 +1,180 @@
+    "org.inviwo.RasterizationRenderer",  // Class identifier
+    "Rasterization Renderer",            // Display name
+    "Mesh Rendering",                    // Category
+    CodeState::Experimental,             // Code state
```
Still experimental?

ResearchDaniel commented on modules/meshrenderinggl/src/processors/rasterizationrenderer.cpp:
```diff
@@ -0,0 +1,180 @@
+
+RasterizationRenderer::IllustrationSettings::IllustrationSettings()
+    : enabled_("illustration", "Use Illustration Effects", false)
+    , edgeColor_("edgeColor", "Edge Color", vec3(0.f, 0.f, 0.f))
```
Could use the new constraints now?
nice to have, not need to...

petersteneteg replied:
use 
`edgeColor_("edgeColor", "Edge Color", ordinal::color(vec3(0.0f)))`

ResearchDaniel replied:
Nice, if time admits it could be worthwhile to go through each property and see if constraints should be added

ResearchDaniel commented on modules/meshrenderinggl/src/rendering/fragmentlistrenderer.cpp:
```diff
@@ -1,562 +1,601 @@
+    // check if enough space was available
+    if (numFrags > fragmentSize_) {
+        // we have to resize the fragment storage buffer
+        LogInfo("fragment lists resolved, pixels drawn: "
```
Remove? Just for debugging?

ResearchDaniel commented on modules/meshrenderinggl/src/rendering/fragmentlistrenderer.cpp:
```diff
@@ -1,562 +1,601 @@
+        // render fragment list
+        display_.activate();
+        setUniforms(display_, textureUnits_[0]);
+        if (builtWithBackground_) {
```
it is just rebuilt above if background exists. Is it possible that the building does not include background anyway? In that case it will be rebuilt every frame?

ResearchDaniel replied:
@AnkeAnke ping

ResearchDaniel commented on modules/meshrenderinggl/src/rendering/fragmentlistrenderer.cpp:
```diff
@@ -1,562 +1,601 @@
+        if (builtWithBackground_) {
+            utilgl::bindAndSetUniforms(display_, textureUnits_, *bg->getData(), "bg",
+                                       ImageType::ColorDepth);
+            display_.setUniform("reciprocalDimensions", vec2(1) / vec2(screenSize_));
```
Is screenSize_ always the same as the background size?
reciprocalDimensions could possibly be renamed to bgReciprocalDimensions to make its intention more explicit

ResearchDaniel replied:
@AnkeAnke ping

ResearchDaniel commented on modules/meshrenderinggl/src/rendering/fragmentlistrenderer.cpp:
```diff
@@ -1,562 +1,601 @@
+        }
+        // reuse fragment lists as neighborhood storage
+
+        LogInfo("Illustration Buffers: additional pixel storage for "
```
Remove

ResearchDaniel commented on modules/meshrenderinggl/src/rendering/fragmentlistrenderer.cpp:
```diff
@@ -1,562 +1,601 @@
+    GLuint v[1] = {0};
+    atomicCounter_.upload(v, sizeof(GLuint));
+    atomicCounter_.unbind();
+    LGL_ERROR;
```
Is there a graceful way of dealing with these possible errors?

ResearchDaniel commented on modules/meshrenderinggl/src/rendering/fragmentlistrenderer.cpp:
```diff
@@ -1,562 +1,601 @@
+        utilgl::bindAndSetUniforms(illustration_.fill, textureUnits_,
+                                   *(backgroundPort_.lock()->getData()), "bg",
+                                   ImageType::ColorDepth);
+        illustration_.fill.setUniform("reciprocalDimensions", vec2(1) / vec2(screenSize_));
```
Same comment as above about background

ResearchDaniel commented on modules/meshrenderinggl/src/rendering/fragmentlistrenderer.cpp:
```diff
@@ -1,562 +1,601 @@
+
+    neighbors.deactivate();
+
+    // perform the bluring
```
blurring

ResearchDaniel commented on modules/meshrenderinggl/src/processors/meshrasterizer.cpp:
```diff
@@ -1,549 +1,500 @@
+
+MeshRasterizer::EdgeSettings::EdgeSettings()
+    : CompositeProperty("edges", "Edges")
+    , edgeThickness_("thickness", "Thickness", 2.f, 0.1f, 10.f, 0.1f)
```
In pixels?

AnkeAnke replied:
Seems so, yes.

ResearchDaniel commented on modules/meshrenderinggl/src/processors/meshrasterizer.cpp:
```diff
@@ -1,549 +1,500 @@
+                   {
+                       {"off", "Off", ShadingMode::Off},
+                       {"phong", "Phong", ShadingMode::Phong},
+                       {"pbr", "PBR", ShadingMode::Pbr},
```
Hmm, there are multiple PBR-shading methods. We might want to be explicit if more alternatives are to be added?

AnkeAnke replied:
Was actually not implemented in the shader... Removed now.

ResearchDaniel commented on modules/meshrenderinggl/src/processors/meshrasterizer.cpp:
```diff
@@ -1,549 +1,500 @@
+                                  std::function<void(Shader&)> setUniformsRenderer) const {
+
+    shader_->activate();
+    if (!shader_->isReady()) return;
```
This should be checked before it is activated?

AnkeAnke replied:
Rather not checked, will throw error, which is what we want.

ResearchDaniel commented on modules/meshrenderinggl/include/modules/meshrenderinggl/processors/meshrasterizer.h:
```diff
@@ -1,306 +1,315 @@
+    virtual bool usesFragmentLists() const override {
+        return !forceOpaque_ && FragmentListRenderer::supportsFragmentLists();
+    }
+    virtual Document getDescription() const override;
```
Is is possible to move this outside of the class? Printing info is not really its reponsibility

petersteneteg replied:
I is, think of it as data. We do the same thing for Image and Volume. since we put it in a port we want to be able to inspect it. We could even add port inspectors for the resterizsers so we could look at the renderings...

AnkeAnke replied:
Renamed now.

ResearchDaniel commented on modules/meshrenderinggl/include/modules/meshrenderinggl/datastructures/rasterization.h:
```diff
@@ -0,0 +1,79 @@
+     * \brief Get data description for the network interface.
+     * @return Specific information about this rasterization type/instance.
+     */
+    virtual Document getDescription() const;
```
Is is possible to move this outside of the class? Printing info is not really its reponsibility

petersteneteg replied:
But it is, se above. how else would you describe what's being rendered?

petersteneteg replied:
but we should name it getInfo to be consistent with the other data structures https://github.com/inviwo/inviwo/blob/6068897cfeb1a10a7c5720595f7c6d50b5c6cd02/include/inviwo/core/datastructures/volume/volume.h#L79

petersteneteg replied:
the info is used here
https://github.com/inviwo/inviwo/blob/6068897cfeb1a10a7c5720595f7c6d50b5c6cd02/include/inviwo/core/ports/datainport.h#L267

the default DataTraits implementation will use
https://github.com/inviwo/inviwo/blob/6068897cfeb1a10a7c5720595f7c6d50b5c6cd02/include/inviwo/core/datastructures/datatraits.h#L103

to get the `getInfo` with help from
https://github.com/inviwo/inviwo/blob/6068897cfeb1a10a7c5720595f7c6d50b5c6cd02/include/inviwo/core/util/introspection.h#L199

So we could add a specialisation to the DataTraits class to do this, but then that specialisation would still need to be able to get the information out of the rasteriser some how...

ResearchDaniel replied:
Yes, I think that it is the responsibility of the DataTrait to provide information. The ones in Volume and other data structures are there due to legacy (we did not have DataTraits at the time)...
Having the getInfo in the class breaks the single responsibility principle

petersteneteg replied:
but then you would have to have a way of exposing what's "inside" rasersize any way. And since what is in it depends on which derived class it is, it has to be a virtual function in the base class. Now what should that function return? I can be lots of different thing in a rasterizer, meshes, textures, shaders, etc? hence it probably has to return some kind of generic string description. hence we get what we have now... The reason for having DataTraits it so we can allow for decorating data with info also when we can't add the getInfo method into the data class, like for a std::vector for example.

ResearchDaniel replied:
One trait per derived class would do it. Anyway, I guess convenience wins here

petersteneteg replied:
You can only specialise for the base class not the derived ones, since the port only know knows about the base. Hence that will not work.

ResearchDaniel replied:
True, marked it as resolved

ResearchDaniel reviewed (dismissed):
Nice! Looks very structured for being so complicated. I have some points that might cause bugs though.

AnkeAnke committed:
Correctly handeling mixture of opaque and OIT

AnkeAnke committed:
Removing debugging code from debugging code

AnkeAnke committed:
Renaming getInfo

petersteneteg commented on modules/meshrenderinggl/src/datastructures/rasterization.cpp:
```diff
@@ -31,7 +31,7 @@
+Document Rasterization::getInfo() const {
     Document doc;
     doc.append("p", "Rasterization functor.");
     return doc;
```
This should describe what kind of rasterisation i.e. fancy/line/sphere etc. could also append the getInfo of the mesh data?

petersteneteg replied:
was meant for the implementation one, think this one can be abstract

AnkeAnke replied:
That's the general rasterization info, so only visible if an inherited class does not override the document. Could make it purely virtual of course, but I think the fallback is fine in case of a very simple rasterizer processor.

petersteneteg commented on modules/meshrenderinggl/include/modules/meshrenderinggl/processors/rasterizationrenderer.h:
```diff
@@ -144,8 +144,7 @@ class IVW_MODULE_MESHRENDERINGGL_API RasterizationRenderer : public Processor {
     ImageOutport outport_;
-
-    BoolProperty forceOpaque_;
+    Image intermediateImage_;
```
This could also be an ImageGL I think. but some of the convenience methods might not exits...

AnkeAnke replied:
Exactly. Is more of a hassle and I have to copy code from the glutils. Hope it's not too much of an overhead?

AnkeAnke committed:
Minor fixes, paper citations

AnkeAnke committed:
Removing PBR setting that does nothing

AnkeAnke committed:
Docu fix

AnkeAnke committed:
Fix

AnkeAnke committed:
Removing extended property settings for now

ResearchDaniel commented on modules/meshrenderinggl/src/processors/rasterizationrenderer.cpp:
```diff
@@ -134,19 +136,24 @@ RasterizationRenderer::IllustrationSettings::getSettings() const {
         }
     }
+
+    // auto inImage = imageInport_->getData();
```
remove

AnkeAnke committed:
Fix

ResearchDaniel commented:
@petersteneteg I just pinged @AnkeAnke about the remaining issues. One of them is a possible shader-rebuild every iteration....

AnkeAnke commented:
The `builtWithBackground_` flag saves whether the shader was build for use with backgrounds. Only if that flag does not agree with the input (i.e., build with background but no background incoming or vice versa), a rebuild is triggered.

ResearchDaniel commented:
Not if !supportsFragmentLists()
Then it will build every frame?

AnkeAnke commented:
That's true. I'll fix it with the next PR.

ResearchDaniel commented:
@petersteneteg please ask before force merging next time :)
